File: benchmarks/wrappers/aws/python/nosql.py
from decimal import Decimal
from os import environ
from typing import List, Optional, Union, Tuple

import boto3


class nosql:

    instance: Optional["nosql"] = None

    def __init__(self):
        self.client = boto3.resource("dynamodb")
        self._tables = {}


    # Based on: https://github.com/boto/boto3/issues/369#issuecomment-157205696
    def _remove_decimals(self, data: dict) -> Union[dict, list, int, float]:

        if isinstance(data, list):
            return [self._remove_decimals(x) for x in data]
        elif isinstance(data, dict):
            return {k: self._remove_decimals(v) for k, v in data.items()}
        elif isinstance(data, Decimal):
            if data.as_integer_ratio()[1] == 1:
                return int(data)
            else:
                return float(data)
        else:
            return data

    def _get_table(self, table_name: str):

        if table_name not in self._tables:

            env_name = f"NOSQL_STORAGE_TABLE_{table_name}"

            if env_name in environ:
                aws_name = environ[env_name]
                self._tables[table_name] = self.client.Table(aws_name)
            else:
                raise RuntimeError(
                    f"Couldn't find an environment variable {env_name} for table {table_name}"
                )

        return self._tables[table_name]

    def insert(
        self,
        table_name: str,
        primary_key: Tuple[str, str],
        secondary_key: Tuple[str, str],
        data: dict,
    ):
        for key in (primary_key, secondary_key):
            data[key[0]] = key[1]

        self._get_table(table_name).put_item(Item=data)

    def get(
        self, table_name: str, primary_key: Tuple[str, str], secondary_key: Tuple[str, str]
    ) -> dict:

        data = {}
        for key in (primary_key, secondary_key):
            data[key[0]] = key[1]

        res = self._get_table(table_name).get_item(Key=data)
        return self._remove_decimals(res["Item"])

    def query(self, table_name: str, primary_key: Tuple[str, str], _: str) -> List[dict]:

        res = self._get_table(table_name).query(
            KeyConditionExpression=f"{primary_key[0]} = :keyvalue",
            ExpressionAttributeValues={":keyvalue": primary_key[1]},
        )["Items"]
        return self._remove_decimals(res)


    def delete(self, table_name: str, primary_key: Tuple[str, str], secondary_key: Tuple[str, str]):
        data = {}
        for key in (primary_key, secondary_key):
            data[key[0]] = key[1]

        self._get_table(table_name).delete_item(Key=data)

    @staticmethod
    def get_instance():
        if nosql.instance is None:
            nosql.instance = nosql()
        return nosql.instance

File: benchmarks/wrappers/aws/nodejs/nosql.js
// Copyright 2020-2025 ETH Zurich and the SeBS authors. All rights reserved.
//
// This is pretty much a Node.js rewrite of our Python wrapper.



const { DynamoDBDocument } = require("@aws-sdk/lib-dynamodb");
const { DynamoDB } = require("@aws-sdk/client-dynamodb");

class nosql {
  constructor() {
    this.client = DynamoDBDocument.from(new DynamoDB());
    this._tables = {};
  }

  _get_table(table_name) {
    if (!(table_name in this._tables)) {
      const env_name = `NOSQL_STORAGE_TABLE_${table_name}`;
      if (env_name in process.env) {
        this._tables[table_name] = process.env[env_name];
      } else {
        throw new Error(
          `Couldn't find an environment variable ${env_name} for table ${table_name}`
        );
      }
    }
    return this._tables[table_name];
  }

  async insert(table_name, primary_key, secondary_key, data) {
    data[primary_key[0]] = primary_key[1];
    data[secondary_key[0]] = secondary_key[1];
    await this.client
      .put({ TableName: this._get_table(table_name), Item: data });
  }

  async get(table_name, primary_key, secondary_key) {
    const key = {};
    key[primary_key[0]] = primary_key[1];
    key[secondary_key[0]] = secondary_key[1];
    const res = await this.client
      .get({ TableName: this._get_table(table_name), Key: key });
    return res.Item;
  }

  async update(table_name, primary_key, secondary_key, updates) {

    const key = {};
    key[primary_key[0]] = primary_key[1];
    key[secondary_key[0]] = secondary_key[1];

    const update_names = {};
    const update_values = {};
    const update_expression = ["SET"];
    for (const [key_name, value] of Object.entries(updates)) {
      update_expression.push(`#${key_name}_name = :${key_name}_value,`);
      update_names[`#${key_name}_name`] = key_name;
      update_values[`:${key_name}_value`] = value;
    }
    // remove trailing comma from the last assignment
    update_expression[update_expression.length - 1] = update_expression[
      update_expression.length - 1
    ].slice(0, -1);

    await this.client
      .update({
        TableName: this._get_table(table_name),
        Key: key,
        UpdateExpression: update_expression.join(" "),
        ExpressionAttributeNames: update_names,
        ExpressionAttributeValues: update_values,
      });
  }

  async query(table_name, primary_key, _) {
    const key_name = primary_key[0];
    const res = await this.client
      .query({
        TableName: this._get_table(table_name),
        KeyConditionExpression: "#key_name = :keyvalue",
        ExpressionAttributeNames: { "#key_name": key_name },
        ExpressionAttributeValues: { ":keyvalue": primary_key[1] },
      });
    return res.Items;
  }

  async delete(table_name, primary_key, secondary_key) {
    const key = {};
    key[primary_key[0]] = primary_key[1];
    key[secondary_key[0]] = secondary_key[1];
    await this.client
      .delete({ TableName: this._get_table(table_name), Key: key });
  }

  static get_instance() {
    if (!nosql.instance) {
      nosql.instance = new nosql();
    }
    return nosql.instance;
  }
}
exports.nosql = nosql;
